README
Vvtbi is copyrighted open-source software that is released under the Apache
License, version 2.0.  For details on the license, see the LICENSE file.

Please send any questions to jahan.addison@jacata.me
    --jahan.addison

CHANGELOG
Changes with vvtbi 1.3
                                                2011-06-20
  *) tokenizer.h: changed token name T_NEWLINE to
      to T_EOL

  *) tokenizer.c (reset): added argument `to' for
      token

  *) vvtbi.c: removed Lines data structure, replaced
      with line-searching algorithm to remove space and
      memory limitation; additionally replacing previous
      (disgusting) fseek method for GOTO and similar pro-
      cedures


Changes with vvtbi 1.2.1
                                                2011-06-10
  *) tokenizer.c (tokenizer_variable_num): added
      better char compatibility

  *) io.c: the `next' character is now stored
      within a variable, replacing fseek -1


Changes with vvtbi 1.2
                                                2011-06-07

  *) io.c (stream): removed function from io.c

  *) main.c: moved open/close logic to io.c

  *) *: all line-statements must now end with a
      new-line character

  *) vvtbi.c (print_statement): fixed unexpected
      behavior that occurs if print is last state-
      ment without trailing new-line

  *) vvtbi.c (variables): changed variable container
      in vvtbi.c from type char[] to int[]


Changes with vvtbi 1.0
                                                2011-05-29

  *) Never released


src/io.c
/********************************
   io.c, @format.new-line  lf
         @format.use-tabs  false
   @format.tab-size    2
   @format.indent-size 2
   @format.line-length 80
*********************************/
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "io.h"

/* I/O file */
static const char *file;

/* I/O FILE stream */
static FILE *Handle = NULL;

/* current characters in stream */
static int current = 0, next = 0;

/******************************************************************************/

/* initialize I/O */

void io_init (const char *fl)
{
  file   = fl;
  Handle = fopen(file, "rb");
  if (!Handle)
  {
    /* failed! */
    fprintf(stderr,
  "*io.c: file `%s' failed!\n",
    file);
    /* exit! */
    exit(EXIT_FAILURE);
  }
  io_next();
}

/* last read character in stream */

int io_current (void)
{
  return current;
}

/* reset current character */

void io_reset (void)
{
  current = 0;
}

/* set to next character in stream */

void io_next (void)
{
  if (!current)
  {
    current = getc(Handle);
    /* save a copy of the next */
    next    = getc(Handle);
  }
  else
  {
  /* set next to current, continue */
    current = next;
    next    = getc(Handle);
  }
}

/* peek the next character in stream */

int io_peek (void)
{
  return next;
}

/* return current position in file */

long io_location (void)
{
  return ftell(Handle);
}

/* set to a new position in file */

void io_seek (long offset, int whence)
{
  fseek(Handle, offset, whence);

}

/* return is stream EOF */

int io_eof (void)
{
  return current == EOF;
}

/* set string dest to n chars in stream */

void to_string (char *dest, size_t n)
{
  size_t i;
  for (i = 0; !io_eof(); i++)
  {
    if (i == n) break;
    dest[i] = io_current();
    io_next();
  }
  dest[i] = 0;
}

/* set to new stream */

void io_set(const char *fl,
  FILE *handle)
{
  if (Handle)
    io_close();
  file   = fl;
  Handle = handle;
}

/* file */

const char *io_file (void)
{
  return file;
}

/* close file stream */

void io_close (void)
{
  fclose(Handle);
}

src/main.c
/*********************************
   main.c, @format.new-line  lf
           @format.use-tabs  false
   @format.tab-size    2
   @format.indent-size 2
   @format.line-length 80
**********************************/
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#define VERSION "1.3"

#include "io.h"
#include "tokenizer.h"
#include "vvtbi.h"

/******************************************************************************/

/*********************/
/* start our program */
/*********************/

int main (int argc, char **argv)
{
  if (argc < 2)
  {
    printf("***\n"
      "@version: " VERSION "\n"
      "********************************\n"
      "  Howto: ./vvtbi[.exe] file\n"
    );
  }
  else
  {
    /**********************/
    /* start interpreter! */
    /**********************/
    vvtbi_init(argv[1]);
    do {
      vvtbi_run();
    } while (!vvtbi_finished());
  }
  /*********/
  /* done! */
  /*********/
  return 0;
}

/*******/
/* End */
/*******/

src/tokenizer.c
/***************************************
   tokenizer.c, @format.new-line  lf
                @format.use-tabs  false
   @format.tab-size    2
   @format.indent-size 2
   @format.line-length 80
****************************************/
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

#include "io.h"
#include "tokenizer.h"

#define MAX_STRING_LEN 50
#define MAX_NUMBER_LEN 6

/* current token */
static int token;

/* scanner's "pointer" */
union Pointer {
  char string[MAX_STRING_LEN+1];
  int number;
  int  letter;
} text;

struct keyword_token {
  char *keyword;
  int   token;
};

/* language keywords */
static const struct keyword_token keywords[] = {
  {"let",   T_LET},
  {"if",    T_IF},
  {"then",  T_THEN},
  {"print", T_PRINT},
  {"rem",   T_REM},
  {"goto",  T_GOTO},
  {NULL,    T_ERROR}
};

static int get_next_token (void);

/******************************************************************************/

/* initialize our file stream and tokenizer */

void tokenizer_init (const char *source)
{
  io_init(source); token = get_next_token();
}

/* a relational or comparison token */

static int token_relation (void)
{
  int next;
  /* peek at the next token */
  next = io_peek();
  switch (io_current())
  {
    case '=':
      io_next();
      return T_EQUAL;
      break;
    case '<':
      io_next();
      if (next == '>')
      {
        io_next();
        return T_NOT_EQUAL;
      }
      else if (next == '=')
      {
        io_next();
        return T_LT_EQ;
      }
      else
      {
        io_next();
        return T_LT;
      }
      break;
    case '>':
      io_next();
      if (next == '=')
      {
        io_next();
        return T_GT_EQ;
      }
      else
      {
        io_next();
        return T_GT;
      }
      break;
  }
  return 0;
}

/* an operation (arithmetic) token */

static int token_operation (void)
{
  switch (io_current())
  {
    case '(':
      return T_LEFT_PAREN;
      break;
    case ')':
      return T_RIGHT_PAREN;
      break;
    case '+':
      return T_PLUS;
      break;
    case '-':
      return T_MINUS;
      break;
    case '/':
      return T_SLASH;
      break;
    case '*':
      return T_ASTERISK;
      break;
  }
  return 0;
}

/* a string token, set our pointer string value */

static int token_string (void)
{
  size_t i = 0;
  io_next(); /* skip current `"' character */
  do
  {
    if (i >= MAX_STRING_LEN) break;
    text.string[i++] = io_current();
    io_next();
  } while(io_current() != '"' &&
    io_current() != EOF);
    io_next();
  if (i > MAX_STRING_LEN)
    text.string[MAX_STRING_LEN] = 0;
  else
    text.string[i] = 0;
  return T_STRING;
}

/* a keyword token */

static int token_keyword (void)
{
  size_t i;
  struct keyword_token const *kt;
  for (i = 0, kt = keywords; kt->keyword; kt++)
  {
    while (toupper(kt->keyword[i]) == io_current())
    {
      i++;
      if (!(kt->keyword[i]))
      {
        /* the REM (comment) statement */
        if (kt->token == T_REM)
        {
          /* skip all characters until newline */
          while(io_current() != '\r'
          && io_current() != '\n'
          && io_current() != EOF)
            io_next();
          return T_REM;
        }
          io_next();
          return kt->token;
      }
      io_next();
    }
  }
  return 0;
}

/* a number (whole number) token, set our Pointer */

static int token_number (void)
{
  int i;
  char number[MAX_NUMBER_LEN+1];
  for (i = 0; i <= MAX_NUMBER_LEN; i++)
  {
    if (!isdigit(io_current()))
    {
      if (i > 0)
      {
        number[i]   = 0;
        text.number = strtol(number, NULL, 10);
        return T_NUMBER;
      }
      else
      {
        break;
      }
    }
    number[i] = io_current();
    io_next();
  }
  return T_ERROR;
}

/* get the next token from the scanner */

static int get_next_token (void)
{
  int c, is_token;
  c = io_current();
  /* EOF token */
  if (c == EOF) return T_EOF;
  /* seperator token */
  if (c == ';' || c == ',')
  {
    io_next(); return T_SEPERATOR;
  }
  /* EOL token */
  if (c == '\r')
  {
    c = io_peek(); io_next();
    if (c == '\n')
    {
      io_next(); return T_EOL;
    }
  }
  else if (c == '\n')
  {
    io_next(); return T_EOL;
  }
  /* relation token */
  is_token = token_relation();
  if (is_token) return is_token;
  /* operation token */
  is_token = token_operation();
  if (is_token)
  {
    io_next(); return is_token;
  }
  /* keyword token */
  is_token = token_keyword();
  if (is_token) return is_token;

  /* string token */
  if (c == '"')   return token_string();
  /* number token */
  if (isdigit(c)) return token_number();
  /* letter token */
  if (isalnum(c) && islower(c))
  {
    text.letter = c;
    io_next();
    return T_LETTER;
  }
  io_next();
  /* N/A */
  return T_ERROR;
}

/* move to the next token in the scanner */

void tokenizer_next (void)
{
  if (tokenizer_finished())
    return;
  /* skip whitespace */
  while (io_current() == ' ' ||
  io_current() == '\t')
    io_next();
  token = get_next_token();
}

/* current token in scanner */

int tokenizer_token (void)
{
  return token;
}

/* string from scanner Pointer */

char *tokenizer_string(void)
{
  return text.string;
}

/* number from scanner Pointer */

int tokenizer_num (void)
{
  return text.number;
}

/* variable number from scanner Pointer */

int tokenizer_variable_num (void)
{
  switch (text.letter)
  {
    case 'a': return 0;  break;
    case 'b': return 1;  break;
    case 'c': return 2;  break;
    case 'd': return 3;  break;
    case 'e': return 4;  break;
    case 'f': return 5;  break;
    case 'g': return 6;  break;
    case 'h': return 7;  break;
    case 'i': return 8;  break;
    case 'j': return 9;  break;
    case 'k': return 10; break;
    case 'l': return 11; break;
    case 'm': return 12; break;
    case 'n': return 13; break;
    case 'o': return 14; break;
    case 'p': return 15; break;
    case 'q': return 16; break;
    case 'r': return 17; break;
    case 's': return 18; break;
    case 't': return 19; break;
    case 'u': return 20; break;
    case 'v': return 21; break;
    case 'w': return 22; break;
    case 'x': return 23; break;
    case 'y': return 24; break;
    case 'z': return 25; break;
  }
  /* will not reach here */
  return 0;
}

/* reset token */

void reset (int to)
{
  token = to;
}

/* scanner is complete */

int tokenizer_finished (void)
{
  if (token == T_EOF)
    /* close I/O */
    io_close();
  return token == T_EOF;
}

src/vvtbi.c
/**********************************
   vvtbi.c, @format.new-line  lf
            @format.use-tabs  false
   @format.tab-size    2
   @format.indent-size 2
   @format.line-length 80
***********************************/

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <ctype.h>

#include "io.h"
#include "tokenizer.h"
#include "vvtbi.h"

#define MAX_STRING_LEN   50
#define MAX_NUMBER_LEN   6
#define MAX_VARIABLE_NUM 26 /* a - z */

/* error constants */
typedef enum {
  E_WARNING,
  E_ERROR
} Error;

/* strings of scanner tokens */
static const char *token_stringify[] = {
  NULL, /* skip 0 index */
  "T_ERROR",
  "T_EOF",
  "T_NUMBER",
  "T_LETTER",
  "T_STRING",
  "T_EQUAL",
  "T_LT",
  "T_GT",
  "T_LT_EQ",
  "T_GT_EQ",
  "T_NOT_EQUAL",
  "T_SEPERATOR",
  "T_MINUS",
  "T_PLUS",
  "T_ASTERISK",
  "T_SLASH",
  "T_LET",
  "T_IF",
  "T_THEN",
  "T_PRINT",
  "T_REM",
  "T_GOTO",
  "T_LEFT_PAREN",
  "T_RIGHT_PAREN",
  "T_EOL"
};

/* variable container */
static int variables[MAX_VARIABLE_NUM];

static int expression (void);
static void line_statement (void);
static void statement (void);

/******************************************************************************/

/* debugging routine */

void dprintf(const char *format,
  Error error, ...)
{
  va_list args;

  va_start(args, error);
  vfprintf(stderr, format, args);
  va_end(args);
  if (error == E_ERROR)
  {
    exit(EXIT_FAILURE);
  }
}

/* initialize tokenizer */

void vvtbi_init (const char *source)
{
  tokenizer_init(source);
  /* initialize variable container */
  memset(variables, 0, MAX_VARIABLE_NUM);
}

/* accept a token from the scanner */

static void accept (int token)
{
  char nearby[10];
  if (token != tokenizer_token())
  {
    to_string(nearby, sizeof nearby);
    dprintf("*vvtbi.c: "
      "unexpected `%s' near `%s',"
      " expected: `%s'\n",
      E_ERROR,
      token_stringify[tokenizer_token()],
      nearby,
      token_stringify[token]);
  }
  tokenizer_next();
}

/* factor necessary components */

static int factor (void)
{
  int r;
  switch (tokenizer_token())
  {
    case T_NUMBER:
      r = tokenizer_num();
      accept(T_NUMBER);
      break;
    case T_LEFT_PAREN:
      accept(T_LEFT_PAREN);
      r = expression();
      accept(T_RIGHT_PAREN);
      break;
    default:
      r = get_variable(
        tokenizer_variable_num()
      );
      accept(T_LETTER);
      break;
  }
  return r;
}

/* parse term */

static int term (void)
{
  int f1, f2, op;
  f1 = factor();
  op = tokenizer_token();
  while (op == T_ASTERISK ||
  op == T_SLASH)
  {
    tokenizer_next();
    f2 = factor();
    switch (op)
    {
      case T_ASTERISK:
        f1 = f1 * f2;
        break;
      case T_SLASH:
        if (f2 == 0)
        {
          /* divide by zero */
          dprintf("*warning: "
            "divide by zero!\n",
            E_WARNING);
          f1 = 0;
        }
        else
        {
          f1 = f1 / f2;
        }
        break;
    }
    op = tokenizer_token();
  }
  return f1;
}

/* parse expression */

static int expression (void)
{
  int t1, t2, op;
  t1 = term();
  op = tokenizer_token();
  while(op == T_PLUS ||
  op == T_MINUS)
  {
    tokenizer_next();
    t2 = term();
    switch (op)
    {
      case T_PLUS:
        t1 = t1 + t2;
        break;
      case T_MINUS:
        t1 = t1 - t2;
        break;
    }
    op = tokenizer_token();
  }
  return t1;
}

/* parse relation */

static int relation (void)
{
  int r1, r2, op;
  r1 = expression();
  op = tokenizer_token();
  while (op == T_EQUAL ||
  op == T_LT ||
  op == T_GT ||
  op == T_LT_EQ ||
  op == T_GT_EQ ||
  op == T_NOT_EQUAL)
  {
    tokenizer_next();
    r2 = expression();
    switch (op)
    {
      case T_EQUAL:
        r1 = r1 == r2;
        break;
      case T_LT:
        r1 = r1 < r2;
        break;
      case T_GT:
        r1 = r1 > r2;
        break;
      case T_LT_EQ:
        r1 = r1 <= r2;
        break;
      case T_GT_EQ:
        r1 = r1 >= r2;
        break;
      case T_NOT_EQUAL:
        r1 = r1 != r2;
        break;
    }
    op = tokenizer_token();
  }
  return r1;
}

/* search file for line number */

static void goto_line (int number)
{
  char linenum[MAX_NUMBER_LEN+1];
  FILE *search;
  size_t i           = 0;
  int num            = 0, c;
  search             = fopen(io_file(),
    "r");
  while(1)
  {
    c                = getc(search);
    if (c == EOF) break;
    /* save line number and verify */
    if (isdigit(c))
    {
      linenum[i++]   = c;
      while ((c      = getc(search)))
      {
        if (!isdigit(c))
          break;
        if (i > sizeof linenum)
          break;
        linenum[i++] = c;
      }
      linenum[i]     = 0;
      num            = strtol(linenum,
        NULL, 0);
      if (number == num)
        break;
    }

    /* skip until EOL */
    while (c != '\n')
      c = getc(search);
    /* start again */
    i                = 0;
  }
  /* matched! */
  if (num == number)
  {
    /* set (close current) */
    io_set(io_file(), search);
    reset(T_NUMBER);
    io_reset(); io_next();
    /* finished! */
    return;
  }
  /* failed! */
  else
  {
    dprintf("*warning: "
      "could not jump to line `%d'\n",
      E_WARNING,
      number
    );
  }
}

/* parse goto statement */

static void goto_statement (void)
{
  int to; accept(T_GOTO);
  to = tokenizer_num();
  accept(T_NUMBER); accept(T_EOL);
  goto_line(to);
}

/* parse print statement */

static void print_statement (void)
{
  accept(T_PRINT);
  do {
    /* begin parsing print statement segments */
    if (tokenizer_token() == T_STRING)
    {
      printf("%s", tokenizer_string());
      tokenizer_next();
    }
    else if (tokenizer_token() == T_SEPERATOR)
    {
      printf(" ");
      tokenizer_next();
    }
    else if (tokenizer_token() == T_LETTER ||
    tokenizer_token() == T_NUMBER ||
    tokenizer_token() == T_LEFT_PAREN)
    {
      printf("%d", expression());
    }
    else
    {
      break;
    }
    /* last character must be newline */
    if (tokenizer_finished())
    {
      accept(T_EOL);
    }
  } while (tokenizer_token() != T_EOL &&
    tokenizer_token() != T_EOF);
  printf("\n");
  tokenizer_next();
}

/* parse if statement */

static void if_statement (void)
{
  int r, to;
  accept(T_IF); r = relation();
  accept(T_THEN); to = tokenizer_num();
  accept(T_NUMBER); accept(T_EOL);
  if (r) goto_line(to);
}

/* parse let statement */

static void let_statement (void)
{
  int var;
  var = tokenizer_variable_num();
  accept(T_LETTER); accept(T_EQUAL);
  set_variable(var, expression());
  accept(T_EOL);
}

/* parse statement */

static void statement (void)
{
  int token;
  char nearby[10];
  token = tokenizer_token();
  switch (token)
  {
    case T_REM:
      tokenizer_next();
      accept(T_EOL);
      break;
    case T_PRINT:
      print_statement();
      break;
    case T_IF:
      if_statement();
      break;
    case T_GOTO:
      goto_statement();
      break;
    case T_LET:
      accept(T_LET);
      /* fall through */
    case T_LETTER:
      let_statement();
      break;
    default:
      /* not implemented! */
      to_string(nearby, sizeof nearby);
      dprintf("*vvtbi.c: statement(): "
        "not implemented near `%s'\n",
        E_ERROR,
        nearby);
      break;
  }
}

/* parse line statement */

static void line_statement (void)
{
  int token;
  /* skip blank lines */
  if (tokenizer_token() == T_EOL)
  {
    do {
      tokenizer_next();
    } while (tokenizer_token() == T_EOL);
  }
  token = tokenizer_token();
  if (token != T_REM)
  {
    accept(T_NUMBER);
  }
  statement();
}

/* parse line statement */

void vvtbi_run (void)
{
  if (tokenizer_finished())
    return;
  line_statement();
}

/* check if scanner is complete */

int vvtbi_finished (void)
{
  return tokenizer_finished();
}

/* set and store variable */

void set_variable (int place, int value)
{
  if (place >= 0 && place <= MAX_VARIABLE_NUM)
    variables[place] = value;
}

/* get variable from container */

int get_variable (int place)
{
  if (place >= 0 && place <= MAX_VARIABLE_NUM)
    return variables[place];
  return 0;
}